packages/dds/merge-tree/src/segmentPropertiesManager.ts

```diff
@@ -45,133 +47,318 @@ export function copyPropertiesAndManager(
+						pending.msnConsensus = computePropertyValue(pending.msnConsensus, [value]);
+					} else {
+						pending.remote.push(value);
 					}
-					this.pendingKeyUpdateCount[key]++;
-				} else if (!shouldModifyKey(key)) {
-					continue;
 				}
+				properties[key] = computePropertyValue(
+					pending.msnConsensus,
+					pending.remote.map((n) => n.data),
+					pending.local.map((n) => n.data),
+				);
+				if (local || pending.local.empty || properties[key] !== previousValue) {
+					deltas[key] = previousValue;
+				}
+			} else {
+				properties[key] = computePropertyValue(previousValue, [value]);
+				deltas[key] = previousValue;
 			}
+		});
+		this.updateMsn(msn);
+		return rtn;
+	}
 
-			const previousValue: unknown = oldProps[key];
-			// The delta should be null if undefined, as that's how we encode delete
-			// eslint-disable-next-line unicorn/no-null
-			deltas[key] = previousValue === undefined ? null : previousValue;
-			if (newValue === null) {
-				// eslint-disable-next-line @typescript-eslint/no-dynamic-delete
-				delete oldProps[key];
+	/**
+	 * Acknowledges property changes.
+	 * This method acknowledges the property changes based on the provided sequence number and operation.
+	 *
+	 * @param seq - The sequence number for the operation.
+	 * @param msn - The minimum sequence number for the operation.
+	 * @param op - The operation containing property changes.
+	 */
+	public ack(seq: number, msn: number, op: PropsOrAdjust): void {
+		for (const [key, value] of opToChanges(op, seq)) {
+			const change = this.changes.get(key);
+			const acked = change?.local?.shift();
+			assert(change !== undefined && acked !== undefined, "must have local change to ack");
+			// we only track remotes if there are adjusts, as only adjusts make application anti-commutative
+			// this will limit the impact of this change to only those using adjusts. Additionally, we only
+			// need to track remotes at all to support emitting the legacy snapshot format, which only sharedstring
+			// uses. when we remove the ability to emit that format, we can remove all remote op tracking
+			if (value.raw !== undefined && change.remote.empty) {
+				change.msnConsensus = computePropertyValue(change.msnConsensus, [value]);
 			} else {
-				// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
-				oldProps[key] = newValue;
+				change.remote.push(value);
 			}
 		}
+		this.updateMsn(msn);
+	}
 
-		return deltas;
+	/**
+	 * Updates the minimum sequence number (msn).
+	 * This method updates the minimum sequence number and removes any changes that have been acknowledged.
+	 *
+	 * @param msn - The minimum sequence number to update.
+	 */
+	public updateMsn(msn: number): void {
+		for (const [key, pending] of this.changes) {
+			pending.msnConsensus = computePropertyValue(
+				pending.msnConsensus,
+				iterateListValues(pending.remote.first, (n) => {
+					if (n.data.seq <= msn) {
+						n.list?.remove(n);
```

Comment:
i don't think `iterateListValues` is robust to in-place manipulation like this. Maybe easiest to do a pattern like

```ts
while (!pending.empty && pending.remote.first.data.seq <= msn) {
	pending.remote.remove(pending.remote.first);
}
```

Reply:
i fixed the function to allow modifcation